rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md

```diff
@@ -0,0 +1,336 @@
+    state: C,
+    stream: Fuse<S>,
+    partitioner: P,
+    timer: Maybe<Sleep>,
+}
+
+impl<S, C, P> Stream for PartitionBatcher<S, C, P>
+where
+    S: Stream<Item = P::Item>,
+    C: BatchConfig<S::Item>,
+    P: Partitioner,
+{
+    type Item = (Prt::Key, C::Batch);
+}
+```
+
+The S3 sink will then be converted to use the new partitioned batcher,
+with the additional behavior change of encoding events before batching
+rather than after.
+
+#### Add a minimum batch size limit to the batcher
+
+Since the multipart object creation API requires each part to have a
+minimum size, this batcher will also have to enforce a minimum size
+before an upload is initiated. The `Batcher` interface will be
+modified to respect this interface and not emit batches until one is
+ready (except at shutdown).
+
+```rust
+trait BatchConfig<T> {
+    …
+    /// Returns true if it the batch is full enough to send.
+    fn is_batch_ready(&self) -> bool;
+}
+
+struct BatcherSettings {
+    …
+    pub size_minimum: usize,
+}
+
+trait SinkBatchSettings {
+    …
+    const MIN_BYTES: Option<usize>;
+}
+
+struct BatchConfig<D: SinkBatchSettings, S = Unmerged> {
+    …
+    pub min_part_bytes: Option<usize>,
+}
+```
+
+#### Modify the S3 sink to use multi-part upload by default
+
+In addition to above, the S3 sink will require one additional
+configuration item to support the new upload mode:
+
+```rust
+struct S3SinkConfig {
+    …
+    pub recover_partials: bool,
+}
+```
+
+The upload system will be modified to upload batches as parts and then
+assembled once the maximum size or timeout is reached.  This will
+require a upload state tracker to contain the data required to
+assemble the parts when the object is completed. There will be one
+saved state for each output partition, and the identifiers must be
+deleted when the final object is created.
+
+```rust
+struct MultipartUploads {
+    uploads: DashMap<String, MultipartUpload>,
+}
+
+/// Multipart upload state data
+struct MultipartUpload {
+    /// The identifier for the upload process.
+    upload_id: String,
+    /// The time after which this batch will be completed.
+    completion_time: DateTime<UTC>,
+    /// A list of uploaded parts.
+    parts: Vec<MultipartUploadPart>,
+}
+
+struct UploadPart {
+    number: i64,
+    e_tag: String,
+}
+```
+
+## Drawbacks
+
+- While this proposal does not change the timing of objects being
+  created in the bucket, it does increase the number of AWS actions
+  required to create an object. This may slightly increase costs for
+  users.
+- This multipart scheme raises the possibility of leaving unusable
+  data in the S3 bucket due to incomplete multipart uploads. This data
+  is recoverable, but only with an extra processing step.
+
+## Prior Art
+
+The [S3 output plugin](https://docs.fluentd.org/output/s3) for fluentd
+uses a disk buffer to store the object before uploading, but uses the
+standard `PutObject` action to upload the object.
+
+## Alternatives
```

Comment:
I think one other possible alternative that is just a slight twist on the above is to use the AWS S3 bucket to track state by writing some metadata object to track partial uploads. This avoids needing to deal with storing Vector state while solving some of the issues mentioned above around scanning for in-flight partial uploads.

Reply:
What additional data would be stored in this external state file than what is already present in the multipart upload state that S3 already provides?

Reply:
Ah, I was thinking of whatever you'd originally planned to include in Vector's local `data_dir` to resume uploads. A sort of primitive form of remote state.

Reply:
Yeah, the state file would pretty much be just what S3 provides with the difference that, since it is local to the node, it would allow for distinguishing between different node's in-progress uploads and avoid clobbering as above. We could store this on S3 with a local node unique identifier, but then we need to store that identifier locally… and still need the `data_dir`.